Add trainable homography warps

SomTambe opened the pull request:
Through this PR, I am adding trainable homography warps. This is mainly focused around adding a `homography` struct which is compatible with the `warp` function in ImageTransformations.jl.

johnnychen94 commented on src/geometry/warp.jl:
```diff
@@ -0,0 +1,34 @@
+    return homography{T}(i)
+end
+
+# adjoint for round(args; kwargs...) since only adjoint for round(args) was defined.
+function ChainRules.rrule(f::typeof(round), x; kwargs...)
+    ȳ = round(x; kwargs...)
+    function pbs(Δ)
+        return (NoTangent(), Δ*zero(x))
+    end
+    return (ȳ, pbs)
+end
```
`round` is applied to transform `tfm` to get a stable forward result, mostly for backward compatibility purpose. For AD purposes, I think we can skip this part.

SomTambe replied:
If we would differentiate the `warp` function even yet, without this adjoint, it would get stuck because Zygote does not detect the kwargs (at some place here they provided it like this `round(x; digits=15)`), and that was causing the problems. It may cause them problems again.

SomTambe replied:
It would still work on the custom warp function we have made.

DhairyaLGandhi commented:
how far are we on getting this merged?

SomTambe commented:
>how far are we on getting this merged?

I am still working on the following issues - 
- [ ] getting the gradient type of `h` to be equal to the type of `h`.
- [ ] getting non zero gradient.

I think it should take some time on this. 
cc. @johnnychen94

SomTambe commented:
I have added adjoints that make `FloatType` inputs work. I have also added tests for them as well as Colorant Types.
The current tests will fail because `Colorant` types do error at points. 
Currently working on resolving that issue.
cc. @johnnychen94 @DhairyaLGandhi

SomTambe commented:
Summary - 
1. Add `homography` struct and methods.
2. Add adjoints for `homography`, Interpolations, `SVector`, `SMatrix`.
3. Add tests for `∂L/∂g`, `∂g/∂(i,j)`, `∂L/∂W`, `FloatTypes` and `Colorant`.

SomTambe commented:
I am closing this PR, and opening a new PR which will be more nicely structured and framed. 
@johnnychen94 @DhairyaLGandhi
